torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
-
-            * state - a dict holding current optimization state. Its content
-                differs between optimizer classes.
-            * param_groups - a dict containing all parameter groups
+            The state of the local optimizer as a :class:`dict`, containing
+            two entries:
+
+            * state
+                a dict holding the current optimization state, where the exact
+                contents depends on the optimizer class.
+            * param_groups
+                a list of parameter groups, each of which is a dict specifying
+                parameters to be optimized and group-specific optimization
+                options.
         """
         return self.optim.state_dict()
 
     def state_dict(self) -> Dict[str, Any]:
         r"""
-        Returns:
-            the last known global optimizer state, which consist of a list of
-            the shards.
+        Returns the last global optimizer state known to this rank.
 
         .. warning:
-            If the state has not been consolidated, this returns a shard's worth,
-            not the global state.
-
-        .. warning:
-            Returning the global state is limited to the replica which was
-            responsible for the consolidation. The state may also not be up to
-            date, depending on when :meth:`consolidate_state_dict` was last called.
+            If the state has not been consolidated to this rank, this raises a
+            runtime error, and even if it has, the state may not be up-to-date,
+            depending on when :meth:`consolidate_state_dict` was last called.
         """
 
-        if len(self._all_states) == 0:
+        if len(self._all_state_dicts) == 0:
             raise RuntimeError(
-                "Optimizer state has not been consolidated on this rank. \
-                Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
+                "Optimizer state has not been consolidated on this rank. "
+                f"Please call `consolidate_state_dict(to={self.rank})` on "
+                "all ranks beforehand if you meant to save the global state."
             )
 
-        # Unify the shard states and the state that pytorch would expect, given the model.
-        # Indexation needs several redirections, since each shard only knows a limited scope of the model
-        # - get the pytorch compliant parameter indexing
+        # Get the possibly-stale global optimizer state that uses global
+        # parameter indexing
         state_dict = super().state_dict()
 
-        # - go through the per-shard states, which are all indexed locally
-        for rank, s in enumerate(self._all_states):
-            # -- match the local indexing and the global partition, update the corresponding saved state globally
-            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
-                local_index_to_param_id = {
```

Comment:
this was for compatibility with pytorch 1.5 (this comes from fairscale, which is compatible with older pytorch, current-2 and at the time this came out it was 1.5). this is not needed anymore  indeed. There were a couple of update PRs for this code which never got reviewed, it's really good that you can spend some time in that